src/Numerics/Mesh/DSS.jl

```diff
@@ -0,0 +1,286 @@
+using CUDA
+
+"""
+    effect_dss3d!(Q::MPIStateArray,
+        grid::DiscontinuousSpectralElementGrid)
+
+This function computes the 3D direct stiffness summation for all variables in the MPIStateArray.
+
+# Fields
+ - `Q`: MPIStateArray
+ - `grid`: Discontinuous Spectral Element Grid
+"""
+function effect_dss3d!(Q::MPIStateArray, grid::DiscontinuousSpectralElementGrid)
+    DA = arraytype(grid)                                      # device array
+    device = arraytype(grid) <: Array ? CPU() : CUDADevice()  # device
+    FT = eltype(Q.data)
+    #----communication--------------------------
+    event = MPIStateArrays.begin_ghost_exchange!(Q)
+    event = MPIStateArrays.end_ghost_exchange!(Q, dependencies = event)
+    wait(event)
+    #----Direct Stiffness Summation-------------
+    nvars = size(Q.data, 2)
+    vertmap = grid.vertmap
+    edgemap = grid.edgemap
+    facemap = grid.facemap
+
+    vtconn = grid.topology.vtconn
+    fcconn = grid.topology.fcconn
+    edgconn = grid.topology.edgconn
+
+    vtconnoff = grid.topology.vtconnoff
+    edgconnoff = grid.topology.edgconnoff
+
+    nvt, nfc, nedg =
+        length(vtconnoff) - 1, size(fcconn, 1), length(edgconnoff) - 1
+
+    Nq = polynomialorders(grid) .+ 1
+    Nqmax = maximum(Nq)
+    Nemax = Nqmax - 2
+    Nfmax = size(facemap, 1)
+
+    if device == CPU()
+        effect_dss3d_CPU!(
+            Q.data,
+            vertmap,
+            edgemap,
+            facemap,
+            vtconn,
+            vtconnoff,
+            edgconn,
+            edgconnoff,
+            fcconn,
+            nvt,
+            nedg,
+            nfc,
+            Nemax,
+            Nfmax,
+        )
+    else
+        max_threads = 256
+        n_items = nvt + nfc + nedg
+        tx = max(n_items, max_threads)
+        bx = cld(n_items, tx)
+        @cuda threads = (tx) blocks = (bx) effect_dss3d_CUDA!(
+            Q.data,
+            vertmap,
+            edgemap,
+            facemap,
+            vtconn,
+            vtconnoff,
+            edgconn,
+            edgconnoff,
+            fcconn,
+            nvt,
+            nedg,
+            nfc,
+            Nemax,
+            Nfmax,
+        )
+    end
```

Comment:
```suggestion
    args = (
      Q.data,
      vertmap,
      edgemap,
      facemap,
      vtconn,
      vtconnoff,
      edgconn,
      edgconnoff,
      fcconn,
      nvt,
      nedg,
      nfc,
      Nemax,
      Nfmax
      )
    if device == CPU()
        effect_dss3d_CPU!(args...)
    else
        max_threads = 256
        n_items = nvt + nfc + nedg
        tx = max(n_items, max_threads)
        bx = cld(n_items, tx)
        @cuda threads = (tx) blocks = (bx) effect_dss3d_CUDA!(args...)
    end
```

Reply:
Would this not result in an additional copy? If so, that would be extremely inefficient. This function is performance critical in the context of continuous Galerkin implemention. A slightly spreadout argument list would be okay in this case.

Reply:
Time it and try? I am not sure the splat should impact performance much.

Reply:
I did not notice much difference in the timing results for the test. I am making the revision shortly.